### ISSUE TITLE
Use of uninitialised variable in bio_print.c:fmtint

### ISSUE THREAD

**Issue by shao-hua-li** [openssl/openssl#17237]
Hi there,

I found an use-of-uninitialized-value when fuzzing `asn1`.

- Configure:
```
CC=clang CFLAGS="-fsanitize=memory -fsanitize-memory-track-origins=2"  ./config enable-fuzz-libfuzzer \
        --with-fuzzer-lib=$PATH_TO_LIBFUZZER \
        -DPEDANTIC no-shared \
        -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION \
        -fsanitize=fuzzer-no-link \
        enable-ec_nistp_64_gcc_128 -fno-sanitize=alignment \
        enable-weak-ssl-ciphers enable-rc5 enable-md2 \
        enable-ssl3 enable-ssl3-method enable-nextprotoneg \
        --debug
```
- Platform: Ubuntu 20.04.4 LTS, x86-64

- Reproduce: run `./fuzz/asn1-test uninitialized_poc`

- POC: [uninitialized_poc.tar.gz](https://github.com/openssl/openssl/files/7676627/uninitialized_poc.tar.gz) (decompress it first)

- A severe consequence of this issue is that `bio_print` may print out different/random outputs for some inputs. The `            ASN1_item_print(bio, o, 4, i, pctx)` function used in `fuzz/asn1.c` has already be affected by this issue. If you compile `openssl` with `gcc -O0` and `gcc -O2`, this function would print different things on the given input. (See the end of this report for reproducing this consequence.)

Memory sanitizer report:
```
==777253==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x8deb08 in fmtint /openssl-fuzz/debug/memory/clang13-default/crypto/bio/bio_print.c:451:28
    #1 0x8d5a22 in _dopr /openssl-fuzz/debug/memory/clang13-default/crypto/bio/bio_print.c:257:22
    #2 0x8d07e5 in BIO_vprintf /openssl-fuzz/debug/memory/clang13-default/crypto/bio/bio_print.c:890:10
    #3 0x8d04f5 in BIO_printf /openssl-fuzz/debug/memory/clang13-default/crypto/bio/bio_print.c:871:11
    #4 0x162de96 in ASN1_buf_print /openssl-fuzz/debug/memory/clang13-default/crypto/asn1/t_pkey.c:36:13
    #5 0x9ef5cd in do_EC_KEY_print /openssl-fuzz/debug/memory/clang13-default/crypto/ec/ec_ameth.c:332:13
    #6 0x4a3928 in FuzzerTestOneInput /openssl-fuzz/debug/memory/clang13-default/fuzz/asn1.c:350:5
    #7 0x1bfa635 in ExecuteFilesOnyByOne /shared/fuzzers/aflpp_diff/aflpp_diff/utils/aflpp_driver/aflpp_driver.c:191:7
    #8 0x1bfa4a5 in main /shared/fuzzers/aflpp_diff/aflpp_diff/utils/aflpp_driver/aflpp_driver.c
    #9 0x7f59a2b290b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/../csu/libc-start.c:308:16
    #10 0x41f03d in _start (/openssl-fuzz/debug/memory/clang13-default/fuzz/asn1+0x41f03d)

  Uninitialized value was created by a heap allocation
    #0 0x44da8d in malloc (/openssl-fuzz/debug/memory/clang13-default/fuzz/asn1+0x44da8d)
    #1 0xdbdf8c in CRYPTO_malloc /openssl-fuzz/debug/memory/clang13-default/crypto/mem.c:184:12
    #2 0xa30715 in EC_KEY_priv2buf /openssl-fuzz/debug/memory/clang13-default/crypto/ec/ec_key.c:967:16
    #3 0x9ef77c in do_EC_KEY_print /openssl-fuzz/debug/memory/clang13-default/crypto/ec/ec_ameth.c:311:19
    #4 0x4a3928 in FuzzerTestOneInput /openssl-fuzz/debug/memory/clang13-default/fuzz/asn1.c:350:5
    #5 0x1bfa635 in ExecuteFilesOnyByOne /shared/fuzzers/aflpp_diff/aflpp_diff/utils/aflpp_driver/aflpp_driver.c:191:7

SUMMARY: MemorySanitizer: use-of-uninitialized-value /openssl-fuzz/debug/memory/clang13-default/crypto/bio/bio_print.c:451:28 in fmtint
Exiting
```

- To reproduce different outputs:
  (1) modify https://github.com/openssl/openssl/blob/f4f77c2d9756cee12875397276799a93f057d412/fuzz/asn1.c#L309 to `BIO *bio = BIO_new_fp(stdout, "w");` so that the output would be printing.
  (2) configure and compile with `gcc -O0` as such (I was using `gcc-11`) :
```
CC=afl-gcc CFLAGS="-O0" ./config enable-fuzz-afl no-shared no-module \
    -DPEDANTIC enable-tls1_3 enable-weak-ssl-ciphers enable-rc5 \
    enable-md2 enable-ssl3 enable-ssl3-method enable-nextprotoneg \
    enable-ec_nistp_64_gcc_128 -fno-sanitize=alignment \
    --debug
```
  (3) configure and compile with `gcc -O2` by replace `CFLAGS="-O0"` to `CFLAGS="-O2"`.
  (4) run `./fuzz/asn1 uninitialized_poc` with differently compiled openssl. You will observe they print whole different things.

**Comment by t8m** [comment]
Could it be that the afl fuzzer somehow disregards the OPENSSL_cleanse() as being setting the memory to 0s? Could you please try this patch?

```
diff --git a/crypto/bn/bn_lib.c b/crypto/bn/bn_lib.c
index 28a3e91679..2a57937c10 100644
--- a/crypto/bn/bn_lib.c
+++ b/crypto/bn/bn_lib.c
@@ -505,7 +505,7 @@ int bn2binpad(const BIGNUM *a, unsigned char *to, int tolen, endianess_t endiane
     /* Swipe through whole available data and don't give away padded zero. */
     atop = a->dmax * BN_BYTES;
     if (atop == 0) {
-        OPENSSL_cleanse(to, tolen);
+        memset(to, '\0', tolen);
         return tolen;
     }
```

**Comment by shao-hua-li** [comment]
> Could it be that the afl fuzzer somehow disregards the OPENSSL_cleanse() as being setting the memory to 0s? Could you please try this patch?
> 
> ```
> diff --git a/crypto/bn/bn_lib.c b/crypto/bn/bn_lib.c
> index 28a3e91679..2a57937c10 100644
> --- a/crypto/bn/bn_lib.c
> +++ b/crypto/bn/bn_lib.c
> @@ -505,7 +505,7 @@ int bn2binpad(const BIGNUM *a, unsigned char *to, int tolen, endianess_t endiane
>      /* Swipe through whole available data and don't give away padded zero. */
>      atop = a->dmax * BN_BYTES;
>      if (atop == 0) {
> -        OPENSSL_cleanse(to, tolen);
> +        memset(to, '\0', tolen);
>          return tolen;
>      }
> ```

Yes, with this patch, the memory sanitizer won't complain.

**Comment by t8m** [comment]
> Yes, with this patch, the memory sanitizer won't complain.

Hmmm... I suppose the patch should be applied as there I do not see any reason why OPENSSL_cleanse should be used there - this is clearing a buffer that is going to be later used by the caller and not a memory to be immediately freed(). I'll create a PR. The question is what is the other issue.

**Comment by t8m** [comment]
> However, the issue that `gcc -O0` and `gcc -O2` compiled binaries output differently still exists.

Could you please paste the different outputs here? I do not have AFL to try? Or is that different output reproducible with normal gcc builds?

**Comment by t8m** [comment]
> Or is that different output reproducible with normal gcc builds?

I am unable to reproduce any difference in the output with normal gcc build with -O2 and with -O0.

**Comment by shao-hua-li** [comment]
Here is the detailed reproduce steps for normal gcc:

(1) modify https://github.com/openssl/openssl/blob/f4f77c2d9756cee12875397276799a93f057d412/fuzz/asn1.c#L309 to `BIO *bio = BIO_new_fp(stdout, "w");` so that the output would be printing.
(2) compile with `gcc -O0` as such 
```
CC=gcc CFLAGS="-O0 '-D__AFL_LOOP(_A)=_A'" ./config enable-fuzz-afl no-shared no-module --debug && make -j
```
(3) run `./fuzz/asn1-test uninitialized_poc` and you'll see the output :
```
ASN1_ANY: SEQUENCE:
    0:d=0  hl=2 l=inf  cons: SEQUENCE
    2:d=1  hl=2 l=   1 prim:  INTEGER           :-12
    5:d=1  hl=2 l=   1 prim:  OCTET STRING      [HEX DUMP]:00
    8:d=1  hl=2 l=inf  cons:  cont [ 0 ]
   10:d=2  hl=2 l=   5 prim:   OBJECT            :secp224r1
   17:d=2  hl=2 l=   0 prim:   EOC
   19:d=1  hl=2 l=   0 prim:  EOC
    ASN1_PRINTABLE: SEQUENCE:
    0:d=0  hl=2 l=inf  cons: SEQUENCE
    2:d=1  hl=2 l=   1 prim:  INTEGER           :-12
    5:d=1  hl=2 l=   1 prim:  OCTET STRING      [HEX DUMP]:00
    8:d=1  hl=2 l=inf  cons:  cont [ 0 ]
   10:d=2  hl=2 l=   5 prim:   OBJECT            :secp224r1
   17:d=2  hl=2 l=   0 prim:   EOC
   19:d=1  hl=2 l=   0 prim:  EOC
    ASN1_SEQUENCE: SEQUENCE:
    0:d=0  hl=2 l=inf  cons: SEQUENCE
    2:d=1  hl=2 l=   1 prim:  INTEGER           :-12
    5:d=1  hl=2 l=   1 prim:  OCTET STRING      [HEX DUMP]:00
    8:d=1  hl=2 l=inf  cons:  cont [ 0 ]
   10:d=2  hl=2 l=   5 prim:   OBJECT            :secp224r1
   17:d=2  hl=2 l=   0 prim:   EOC
   19:d=1  hl=2 l=   0 prim:  EOC
    SEQUENCE OF ASN1_SEQUENCE_ANY {
      INTEGER:-18

      OCTET STRING:
        0000 - 00                                             .

      (unknown):
    0:d=0  hl=2 l=inf  cons: cont [ 0 ]
    2:d=1  hl=2 l=   5 prim:  OBJECT            :secp224r1
    9:d=1  hl=2 l=   0 prim:  EOC
    }
Private-Key: (224 bit)
priv:
    00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:
    00:00:00:00:00:00:00:00:00:00:00:00:00
pub:
    00
ASN1 OID: secp224r1
NIST CURVE: P-224
Private-Key: (224 bit)
priv:
    00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:
    00:00:00:00:00:00:00:00:00:00:00:00:00
pub:
    00
ASN1 OID: secp224r1
NIST CURVE: P-224
```

(4) compile again (or make another clone) with `gcc -O2` as such:
```
CC=gcc CFLAGS="-O2 '-D__AFL_LOOP(_A)=_A'" ./config enable-fuzz-afl no-shared no-module --debug && make -j
```
(5) run again `./fuzz/asn1-test uninitialized_poc` and you'll see a much shorter output:
```
ASN1_ANY: SEQUENCE:
    0:d=0  hl=2 l=inf  cons: SEQUENCE
    2:d=1  hl=2 l=   1 prim:  INTEGER           :-12
    5:d=1  hl=2 l=   1 prim:  OCTET STRING      [HEX DUMP]:00
    8:d=1  hl=2 l=inf  cons:  cont [ 0 ]
   10:d=2  hl=2 l=   5 prim:   OBJECT            :secp224r1
   17:d=2  hl=2 l=   0 prim:   EOC
   19:d=1  hl=2 l=   0 prim:  EOC
```

**Comment by bernd-edlinger** [comment]
> BIO *bio = BIO_new_fp(stdout, "w");

this is actually wrong, and causes undefined behavour,
the signature of BIO_new_fp isi
`BIO *BIO_new_fp(FILE *stream, int close_flag);`

so "w" is not zero, and causes this function to close stdout, therefore
in the output is just already closed with -O2 after the first match,
the longer output with -O0 is more correct, but in this involves undefined
behaviour, therefore the compiler is free to ignore that close attempt or not.

Try `BIO *bio = BIO_new_fp(stdout, 0);` instead.

**Comment by shao-hua-li** [comment]
> > BIO *bio = BIO_new_fp(stdout, "w");
> 
> this is actually wrong, and causes undefined behaviour, the signature of BIO_new_fp is `BIO *BIO_new_fp(FILE *stream, int close_flag);`
> 
> so "w" is not zero, and causes this function to close stdout, therefore the output stream is just already closed with -O2 after the first match, the longer output with -O0 is more correct, but as this involves undefined behaviour, the compiler is free to ignore that close attempt or not.
> 
> Try `BIO *bio = BIO_new_fp(stdout, 0);` instead.

Thanks for the answer. Yes, now all compilers/opts are printing the same thing.

**Comment by t8m** [comment]
So there is only the initial false positive error that is resolved by #17240
